src/map/pc.cpp

```diff
@@ -12058,253 +12169,544 @@ static unsigned int pc_calc_basesp(uint16 level, uint16 class_) {
-	for(i = 0; i <= MAX_WEAPON_TYPE; i++)
-#else
-	for(i = 0; i < MAX_WEAPON_TYPE; i++)
-#endif
-	{
-		job_info[idx].aspd_base[i] = atoi(fields[i+5]);
+	if (!script_get_constant(job_name_constant.c_str(), &job_id)) {
+		this->invalidWarning(node["Job"], "Job %s does not exist.\n", job_name.c_str());
+		return 0;
 	}
 
-	return true;
-}
+	std::shared_ptr<s_job_info> job = this->find(static_cast<uint16>(job_id));
+	bool exists = job != nullptr;
 
-//Reading job_db2.txt line (class,JobLv1,JobLv2,JobLv3,...)
-static bool pc_readdb_job2(char* fields[], int columns, int current)
-{
-	int idx, class_, i;
+	if (!exists)
+		job = std::make_shared<s_job_info>();
 
-	class_ = atoi(fields[0]);
+	if (this->nodeExists(node, "MaxWeight")) {
+		int32 weight;
 
-	if(!pcdb_checkid(class_))
-	{
-		ShowWarning("status_readdb_job2: Invalid job class %d specified.\n", class_);
-		return false;
-	}
-	idx = pc_class2idx(class_);
+		if (!this->asInt32(node, "MaxWeight", weight))
+			return 0;
 
-	for(i = 1; i < columns; i++)
-	{
-		job_info[idx].job_bonus[i-1] = atoi(fields[i]);
+		job->max_weight_base = weight;
+	} else {
+		if (!exists)
+			job->max_weight_base = 20000;
 	}
-	return true;
-}
 
-//Reading job_exp.txt line
-//Max Level,Class list,Type (0 - Base Exp; 1 - Job Exp),Exp/lvl...
-static bool pc_readdb_job_exp(char* fields[], int columns, int current)
-{
-	int idx, i, type;
-	int job_id,job_count,jobs[CLASS_COUNT];
-	unsigned int ui, maxlvl;
+	if (this->nodeExists(node, "HPFactor")) {
+		int32 hp;
 
-	maxlvl = atoi(fields[0]);
-	if(maxlvl > MAX_LEVEL || maxlvl<1){
-		ShowError("pc_readdb_job_exp: Invalid maxlevel %d specified.\n", maxlvl);
-		return false;
+		if (!this->asInt32(node, "HPFactor", hp))
+			return 0;
+
+		job->hp_factor = hp;
+	} else {
+		if (!exists)
+			job->hp_factor = 20000;
 	}
-	if((maxlvl+3) > columns){ //nb values = (maxlvl-startlvl)+1-index1stvalue
-		ShowError("pc_readdb_job_exp: Number of columns %d defined is too low for max level %d.\n",columns,maxlvl);
-		return false;
+
+	if (this->nodeExists(node, "HPMultiplicator")) {
+		int32 hp;
+
+		if (!this->asInt32(node, "HPMultiplicator", hp))
+			return 0;
+
+		job->hp_multiplicator = hp;
+	} else {
+		if (!exists)
+			job->hp_multiplicator = 500;
 	}
-	type = atoi(fields[2]);
-	if(type < 0 || type > 1){
-		ShowError("pc_readdb_job_exp: Invalid type %d specified.\n", type);
-		return false;
+
+	if (this->nodeExists(node, "SPFactor")) {
+		int32 sp;
+
+		if (!this->asInt32(node, "SPFactor", sp))
+			return 0;
+
+		job->sp_factor = sp;
+	} else {
+		if (!exists)
+			job->sp_factor = 100;
 	}
-	job_count = pc_split_atoi(fields[1],jobs,':',CLASS_COUNT);
-	if (job_count < 1)
-		return false;
-	job_id = jobs[0];
-	if(!pcdb_checkid(job_id)){
-		ShowError("pc_readdb_job_exp: Invalid job class %d specified.\n", job_id);
-		return false;
+
+	if (this->nodeExists(node, "BaseASPD")) {
+		const YAML::Node &aspdNode = node["BaseASPD"];
+		bool shield =
+#ifdef RENEWAL
+			true;
+#else
+			false;
+#endif
```

Comment:
Maybe calculate "maximum" here instead.
You only need the shield thing for calculating the maximum multiple times later.